### PR TITLE
Remove Action monad, add focus operator, Semigroup and Monoid instances

### PR DESCRIPTION
**Pull request by paf31**
@ethul @kRITZCREEK What do you think about this?

This adds `Monoid` and a lens/prism operator for `Spec`, so that there is a means to compose components. I based this on `optic-ui` since I think it's the most solid form of component composition I've seen.

The next step would be to add a `foreach` operator to create things like list components using a `Traversal`, as in `optic-ui` or `syb`.

/cc @zrho

### CODE REVIEW

**Comment by jasonzoladz** [comment]
Continual bike-shedding of this api hurts usability.  :(

**Comment by paf31** [comment]
@jasonzoladz You might be right, but I'm definitely not happy with certain things in the API right now. Specifically, component composition is harder than it needs to be, and the `Action` monad is a constant source of confusion/frustration. I'm going to leave this one open for a while, so people can try it out.

When I'm happy with the API, I'll call it 1.0, and breaking changes will slow down considerably, but you should expect some breakage until we get there.

**Comment by ethul** [comment]
I think this looks great! Although I admit I am still a lens newbie. Do you have an example or gist showing this in action?

**Comment by jasonzoladz** [comment]
@paf31 Fair enough.  If you're considering more drastic changes you might look at https://hackage.haskell.org/package/react-flux for inspiration.  It's really nice.

**Comment by paf31** [comment]
@ethul Not yet, but hopefully I will soon.

**Comment by paf31** [comment]
@ethul I added an example, but I still need to figure out some stuff with the React effects.

**Comment by ethul** [comment]
Thanks for the example!

File: test/Main.purs [hunk 1]
```diff
@@ -42,24 +45,30 @@ render dispatch s _ _ = R.div' [counter, buttons]
   buttons :: R.ReactElement
   buttons =
     R.p'
-      [ R.button [ RP.onClick (\_ -> dispatch Increment) ]
+      [ R.button [ RP.onClick (\_ -> unsafeInterleaveEff $ dispatch Increment) ]
```
Comment (passy) [inline test/Main.purs]: Is this supposed to remain like this?

Comment (paf31) [inline test/Main.purs]: No, I still need to figure out how to handle the React effects properly.

**Comment by paf31** [comment]
Ok, this is ready for review. @ethul @jasonzoladz I'd be interested to get any other feedback you might have.

**Comment by passy** [comment]
I'm a fan of those changes. I've updated the todomvc example for this branch: https://github.com/functorial/purescript-thermite-todomvc/compare/gh-pages...passy:updates

(Not sure if this is the correct way of doing it, though ...)

**Comment by paf31** [comment]
@passy Looks nice! I was actually working on something similar as a new test app, which I've added to this PR now. It uses a separate component for tasks to demonstrate the lensy composition of components.

The difference in composition for `ReactElement` and `Spec` is something I'd like to try to tackle at some point in the future, but for now, I think the composition is probably lightweight enough to not be a big problem.

One remaining possible issue is that actions can currently only be handled by at most one `PerformAction` function, due to the way `setState` gets called. This might be something to revisit later.

**Comment by paf31** [comment]
Still to do after a bit of thought:

- [ ] Figure out what to do with `componentWillMount` etc.
- [ ] Tidy up the effects

**Comment by paf31** [comment]
Does anyone have any thoughts on what to do with `componentWillMount` etc.?

Since Thermite uses `purescript-react` now, and this branch already contains a function to create a `ReactSpec` instead of a component class, I'm leaning towards removing it and giving the user access to the event handler function (what Halogen would call the driver function). This means the user gets to add their own lifecycle methods, and Thermite is just a layer on top which adds the lens operators and action abstraction. Child components can still provide their own initializers by sharing an initialization action with their parents via a prism.

**Comment by ethul** [comment]
I think these changes are looking great. I really like the lens/action abstraction layer idea. For what it's worth, I think opening up the lifecycle methods would be useful and allow more flexibility in the components.

**Comment by paf31** [comment]
@ethul By "opening up the lifecycle methods", are you referring to using the ones from `purescript-react`? if so, that should be possible now, by compiling the `Spec` only as far as a `ReactSpec`, not all the way to a `ComponentClass`.

**Comment by paf31** [comment]
Ok, I've 

- Moved `componentWillMount` out
- Hidden the `Spec` constructor and exposed lenses for its fields (should help with wrapping existing components to add extra elements)
- Tidied up the effects a bit

If we think this API looks okay, I'll work on documentation, and release this as 0.11. Since the implementation of `Spec` is now abstract, I'll be able to work on hiding bits of the state with an existential type internally in 0.12 (/cc @born2defy).

Any thoughts on this before it gets merged?

**Comment by born2defy** [comment]
I was checking it out a bit late last night and already am a big fan of the new lenses into the Spec functions.  I will play with it a bit more today and let you know my thoughts.  

Adam


On Oct 25, 2015, at 10:58 PM, Phil Freeman <notifications@github.com> wrote:

Ok, I've

Moved componentWillMount out
Hidden the Spec constructor and exposed lenses for its fields (should help with wrapping existing components to add extra elements)
Tidied up the effects a bit
If we think this API looks okay, I'll work on documentation, and release this as 0.11. Since the implementation of Spec is now abstract, I'll be able to work on hiding bits of the state with an existential type internally in 0.12 (/cc @born2defy).

Any thoughts on this before it gets merged?

**Comment by born2defy** [comment]
One thing occurred to me today.  What are we doing with `props` ?  Without any `lens` love, how do we compose two components who have different `prop` configurations?  Considering how react is structured, it's probably not the best idea to eschew props for state completely.  Having all the props for a complex app in one record split up with `lenses` may be a bit overwhelming - then again it would make it incredibly simple to reconfigure components.  Thoughts?

**Comment by paf31** [comment]
I've thought about `props` quite a bit but it's not clear to me how they relate to the current problem of building a single React component out of parts with possibly shared bits of state. I'm open to suggestions though.

**Comment by ethul** [comment]
@paf31 You're right. Thanks. I will have to dig more into the new release. The changes look great!

**Pull request merged**